WIP [RN] Recording on mobile

zbettenbuk opened the pull request:
The first two commits implement/unify notifications and labels on web and mobile, then the last one implements recording (also unifying the code with web) on mobile, using the previously implemented features.

virtuacoplenny commented on react/features/recording/middleware.js:
```diff
@@ -81,15 +85,19 @@ MiddlewareRegistry.register(store => next => action => {
                 store.dispatch(playSound(RECORDING_OFF_SOUND_ID));
+            } else if (updatedSessionData.status === PENDING
+                && (!oldSessionData || oldSessionData.status !== PENDING)) {
+                store.dispatch(showPendingRecordingMessage());
```
I think it'd be good actually to take into account error states right now too. A recording can go from on to error. I'm not sure if I ever tested pending to error. If a recording fails to start, there is no recording session so the error is handle elsewhere.

virtuacoplenny commented on react/features/recording/components/Recording/RecordButton.native.js:
```diff
@@ -0,0 +1,138 @@
+ * @param {Object} state - The Redux state.
+ * @private
+ * @returns {{
+ *     _isRecordingRunning: boolean,
```
Probably should update the return params here.

virtuacoplenny commented on react/features/conference-messages/reducer.js:
```diff
@@ -0,0 +1,44 @@
+    default:
+        return state;
+    }
+});
```
I remember now (great timing, right?) I was also thinking of asking to get the web recording messages redesigned. Maybe on web it would be good to leverage the notifications feature for recording messages. ConferenceMessages sound a lot like Notifications to me. I never asked...

zbettenbuk replied:
That is a very good point, and I'm working on it right now. The feature "notifications" will be unified over web and mobile and both platform will use them for recording messages

virtuacoplenny commented on react/features/conference-messages/actions.js:
```diff
@@ -0,0 +1,48 @@
+            setTimeout(() => dispatch({
+                type: HIDE_CONFERENCE_MESSAGE,
+                messageKey
+            }), timeout);
```
I think this works well enough for now. One edge case comes to mind and I don't know if it warrants any further thought. Do you think the case could happen where the messageKey gets queued at say 10minutes in the call with a 10 second timeout, then the same message gets queued at 10 minutes and 9 seconds? Right now I can't think of that causing a problem.

zbettenbuk replied:
Yes, that can totally cause issues, and with this, I just realized that I did the same mistake I constantly fight against: no premature features. I removed the entire timing from the messages feature, because even if I implement it properly I can't test it as I don't use it. Someone will implement it with ease when it gets used. Thanks!

zbettenbuk commented:
@lyubomir this PR is pretty much done, except with the recording related error messages that I still ought to implement/test (I'm talking to Aaron about that). So you (or someone who you assign) can start to review it, focusing more on the approach (that is not planned to change), less on the actual coding style (that can still slightly improve).

virtuacoplenny commented on react/features/toolbox/components/native/OverflowMenu.js:
```diff
@@ -68,6 +69,7 @@ class OverflowMenu extends Component<Props> {
                 <ToggleCameraButton { ...buttonProps } />
                 <AudioOnlyButton { ...buttonProps } />
                 <RoomLockButton { ...buttonProps } />
+                <RecordButton { ...buttonProps } />
```
I think the presence of the button will have to be determined by some settings, like moderator and some other config values. Might be good to sync up with Damyan about it to get the details.

zbettenbuk replied:
Very true! But that decision is in the RecordButton, based on _mapStateToProps

virtuacoplenny commented on react/features/recording/middleware.js:
```diff
@@ -76,25 +84,48 @@ MiddlewareRegistry.register(store => next => action => {
     }
+
+    case SHOW_NOTIFICATION:
+        if (action.props.descriptionKey === 'recording.pending') {
```
I wonder if it'll be better to sniff this out some other way, maybe like with an explicit flag. In case someone changes the string that displays and misses this reference.

zbettenbuk replied:
Okay, I check this, thanks!

virtuacoplenny commented on react/features/recording/components/RecordingLabel.web.js:
```diff
@@ -5,243 +5,35 @@ import { connect } from 'react-redux';
-            break;
-        }
-
-        const className = `recording-label ${
```
The amount of appreciation I have for you cleaning the label shenanigans is immeasurable. The label pulling double duty has been an issue for so long...and I've kicked the can down the road twice now, once for vertical filmstrip and once for simultaneously live streaming and recording. You're a better man than I.

zbettenbuk replied:
Nah, I just had enough time to work on it

virtuacoplenny commented on react/features/recording/components/RecordingLabel.native.js:
```diff
@@ -6,52 +6,34 @@ import { connect } from 'react-redux';
-            labelKey = 'recording.live';
             indicatorStyle = styles.indicatorLive;
             break;
         case JitsiRecordingConstants.mode.FILE:
```
Do you think a future iteration of recording should split the label into file and stream labels?

zbettenbuk replied:
It is split now, the logic is in the abstract class. Or is it what you meant?

virtuacoplenny commented on react/features/recording/components/Recording/RecordButton.native.js:
```diff
@@ -0,0 +1,141 @@
+     * @private
+     * @returns {void}
+     */
+    _onStartRecording() {
```
Do you think it's too nested to have an abstract recording button? It might be.

zbettenbuk replied:
It seemed too nested for me (at least for now) as those buttons are handled totally differently on web (or on mobile, depends on who you ask) at the moment. But I'm happy to start a bigger refactor where we unify those codes too (overlay menu...etc)

paweldomas commented on react/features/large-video/components/styles.js:
```diff
@@ -1,13 +1,36 @@
+     * Indicator container for wide aspect ratio.
+     */
+    indicatorContainerWide: {
+        right: 80
```
if this value is not relative to anything please declare a constant for it like the AVATAR_SIZE above

paweldomas commented:
I gave it a try on Android, but there's some problem with cleaning up the state:
1. Enter room A on meet.jit.si
2. Start the recording
3. Jibri joins and "REC" label is displayed.
4. Leave the room with hang up
5. Join room B
At this point the REC label is still displayed and the app thinks that the recording is in progress... The stop recording option will not work

zbettenbuk commented on react/features/recording/reducer.js:
```diff
@@ -11,13 +15,25 @@ const DEFAULT_STATE = {
+                ...state,
+                pendingNotificationUid: action.uid
+            };
+
```
Live streaming was not in my plans, but just got a question from Lyubo indicating that it'll be necessary too, so I'll revisit this

zbettenbuk commented on react/features/recording/actionTypes.js:
```diff
@@ -1,3 +1,17 @@
+ * }
+ * @public
+ */
+export const CLEAR_PENDING_RECORDING_NOTIFICATION_UID
```
Good point, thanks! I'm changing this together with the other notification related refactor

zbettenbuk commented on react/features/notifications/middleware.js:
```diff
@@ -0,0 +1,22 @@
+    switch (action.type) {
+    case CONFERENCE_FAILED:
+    case CONFERENCE_LEFT:
+        dispatch(clearNotifications());
```
Thanks! I didn't know this StateListenerRegistry, but I like the concept, so I try to implement your suggestion.

zbettenbuk commented on react/features/notifications/components/NotificationsContainer.native.js:
```diff
@@ -0,0 +1,145 @@
+    return {
+        ..._abstractMapStateToProps(state),
+        _filmstripVisible:
+            state['features/filmstrip'].visible && participantCount > 1,
```
Yes, thanks!

zbettenbuk commented on react/features/notifications/components/NotificationsContainer.native.js:
```diff
@@ -0,0 +1,145 @@
+
+        // The container needs additional distance from bottom when the
+        // filmstrip or the toolbox is visible.
+        _filmstripVisible && _narrow && (bottom += 100);
```
Ok, thanks!

zbettenbuk commented on react/features/recording/components/Recording/RecordButton.native.js:
```diff
@@ -0,0 +1,141 @@
+    _onStartRecording() {
+        sendAnalytics(createRecordingDialogEvent('start', 'confirm.button'));
+
+        this.props._conference.startRecording({
```
Sh*t, I think I just left it there after a refactor, thanks!

saghul commented on react/features/filmstrip/functions.native.js:
```diff
@@ -0,0 +1,20 @@
+ */
+export function isFilmstripVisible(stateful: Object | Function) {
+    const state = toState(stateful);
+    const { length: participantCount } = state['features/base/participants'];
```
use getParticipantCount here. Generally, we started a transition away from checking `features/base/participants` directly, because one might be accessing stalled content.

zbettenbuk replied:
We've already discussed this, that gives a different result as it excludes bots. I'm okay to change this if someone confirms that bots are okay to be excluded from the count that controls the filmstrip behaviour

paweldomas replied:
I suppose that bots should be excluded from the count if they are not displayed in the filmstrip. Are they ?

paweldomas replied:
Btw. I have specifically asked to have this method introduced in a separate commit. Could you please separate it ?

saghul commented on react/features/notifications/components/AbstractNotification.js:
```diff
@@ -0,0 +1,150 @@
+
+        const descriptionArray = [];
+
+        if (descriptionKey) {
```
use &&

saghul commented on react/features/notifications/components/AbstractNotification.js:
```diff
@@ -0,0 +1,150 @@
+            descriptionArray.push(t(descriptionKey, descriptionArguments));
+        }
+
+        if (description) {
```
ditto

saghul commented:
@zbettenbuk (but also anyone reading this) -- I left a couple of drive-by comments. I just took a look at the notifications part. Here are some thoughts.

First, here is how they look:

<img width="545" alt="screen shot 2018-06-22 at 15 02 10" src="https://user-images.githubusercontent.com/317464/41778086-d29b284a-762d-11e8-8e2c-461e23ce18c5.png">

- The container is not centered
- Why can't the user dismiss this particular notification?
- They show up in the middle of the screen, which is (subjectively) not great UI

Since we are knee-deep in Material-Design style stuff, I'd go for something like: https://github.com/cooperka/react-native-snackbar

There is also a case to be made of notifications out of the conference. Are those useful? I'd say yes. For instance we could show the last failure if the user cancels the error overlay.

In general I don't think notifications can be ported as-is to mobile. Web has a lot more screen real state. If too many unimportant notifications pile up (think participants joining and leaving) we may want to coalesce those and have important ones such as recording to take precedence.

Last, mobile currently has arguably another type of notifications: the calendar ones. Some unification here would be nice.

IMHO notifications require / deserve special attention. Can't recording land without them? There is an audible and visual indication already.

zbettenbuk commented:
@saghul @lyubomir @paweldomas @virtuacoplenny : Yes, notifications can be a good topic for a discussion.

However what came to my mind reading about this suggestion was : "Oh, yet another bloated lib". We can use that for sure, but then we diverge even further from web, which we don't want, as far as I know. There are cases where we need to, of course - but notifications on mobile now uses exactly the same code and logic that web uses, I think this is a win.

And about the questions:
- They are not aligned to the center because I followed web's approach here: bottom left, aligned to left (so as the text inside) - we're building a brand after all
- This one is not dismiss-able because the recording status message was also not dismiss-able originally
- I think they show up at the bottom of the screen, excluding the toolbox and filmstrip. Or what do you mean there?

Good point about the calendar notification though, I think that should also be a simple (dismiss-able) notification.

paweldomas commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+    // simplified.
+    const {
+        connectionEstablished,
+        haveToReload,
```
hmm I don't see neither 'haveToReload' nor 'connectionEstablished' defined on the current master... what am I missing ?

paweldomas replied:
I don't see 'haveToReload' nor 'connectionEstablished' fields here:
https://github.com/jitsi/jitsi-meet/blob/master/react/features/overlay/reducer.js
Was this added as part of this PR or is there magic somewhere ?

paweldomas commented:
> I gave it a try on Android, but there's some problem with cleaning up the state:
>
>  1. Enter room A on meet.jit.si
>  2. Start the recording
>  3. Jibri joins and "REC" label is displayed.
>  4. Leave the room with hang up
>  5. Join room B
>
>At this point the REC label is still displayed and the app thinks that the recording is in progress... The stop recording option will not work

How about this issue ? Did you get a chance to look into that ? Do you reproduce it ?

paweldomas commented:
> @saghul @lyubomir @paweldomas @virtuacoplenny : Yes, notifications can be a good topic for a discussion.

I agree we can start a discussion about it, but as far as this PR goes it should not be blocked by that.

zbettenbuk commented:
> How about this issue ? Did you get a chance to look into that ? Do you reproduce it ?

Sorry I forgot to reply (I hate why I can't reply to comments), yes, thanks for the report, that bug was indeed there and I fixed it with the StateListenerRegistry, similar to how you suggested about the notifications. Btw can you please check those usages, I'm not fully familiar with this new Listener. They are in recording middleware and notification middleware.

paweldomas commented:
> Sorry I forgot to reply (I hate why I can't reply to comments), yes, thanks for the report, that bug was indeed there and I fixed it with the StateListenerRegistry, similar to how you suggested about the notifications. Btw can you please check those usages, I'm not fully familiar with this new Listener. They are in recording middleware and notification middleware.

sure !

paweldomas commented:
> IMHO it should, iff we want recording to land with notifications. We could strip those and land recording then discuss and work on notifications.

> I don't fancy the idea of landing something that we already know is not they way we want it to be. (unless there is time pressure we don't control)

For me this looks totally fine and could stay until we come up with something better, but it's just my opinion. I approve it this way. Lets check with @lyubomir too and then decide whether to strip or not.

paweldomas commented:
![screenshot_20180625-150807](https://user-images.githubusercontent.com/2965063/41857870-be55a596-7898-11e8-8831-1ae4bec1148a.jpg)
There's also this warning. Do you have any idea what could be causing it ?

paweldomas commented:
![screenshot_20180625-155138](https://user-images.githubusercontent.com/2965063/41857948-eeca29ea-7898-11e8-8ddf-b5003c3f9f9d.jpg)
"All recorders are currently busy" - appeared too wide in the portrait mode and close button was not visible.

paweldomas commented:
Also we should probably add "live streaming is starting..." and "live streaming stopped" notifications similar to what we have on web. Starting live streaming is long process and it's good to let the user know that someone is starting it. There are no sounds currently, so it can be a surprise.

paweldomas commented:
Note that I have not tested the web part yet, so I may post more feedback once I do that.

paweldomas commented:
> Also I have a concern about enabling FILE recordings on mobile at this point given that the recording's file is not accessible yet. It would only confuse users. In addition to that I learned from Brian that the file is currently stored on Jibri and without the script which uploads and deletes it, it is possible to fill up Jibri machine. Can we add a flag which would be false by default and the recording menu item would stay hidden until the FILE recording is usable ?

So it appears that this PR is not using the right config option for enabling the FILE recording. Could you please make sure to enable it on `fileRecordingsEnabled` equals `true` ? Currently on meet.jit.si it's disabled, but the mobile app turns it on for me.

zbettenbuk commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+ * }}
+ */
+export function _abstractMapStateToProps(state: Object) {
+    // TODO: Per existing behavior, notifications should not display when an
```
This code was part of the notification code originally, I just relocated it. Are you sure refactoring this logic is part of this PR?

paweldomas replied:
You're right that this is a copy paste. It appears that the old code was broken, so I don't know about refactoring, but IMHO it should be fixed at least. It refers to not existing properties.

zbettenbuk replied:
@virtuacoplenny I'm about to refactor this, but it only affects web. What do you think about this? Can you please give your input what to take care about?

zbettenbuk commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+    // overlay is visible. This logic for checking overlay display can likely be
+    // simplified.
+    const {
+        connectionEstablished,
```
Same as above. I'm happy to refactor, but this is not new code and it's web

zbettenbuk commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+    componentDidUpdate(prevProps: P) {
+        const { _notifications, _showNotifications } = this.props;
+
+        if (_notifications.length) {
```
Good point, thanks! Although, this is also old code, but it was an easy fix, so I changed it.

paweldomas replied:
Yes, it was in old code. Thanks for fixing it !

zbettenbuk commented on react/features/notifications/components/Notification.native.js:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Implements a React {@link Component} to display a notification.
+ *
+ * NOTE: Special notifications that web has, such as error or warning are not
```
Sorry, that comment is a leftover, I removed it.

zbettenbuk commented on react/features/recording/actions.js:
```diff
@@ -1,4 +1,109 @@
+ * @returns {Function}
+ */
+export function showStoppedRecordingNotification() {
+    return (dispatch: Function) => dispatch(showNotification({
```
Thanks, fixed it

zbettenbuk commented on react/features/recording/actions.js:
```diff
@@ -1,4 +1,109 @@
+ * @returns {Function}
+ */
+export function showRecordingError(props: Object) {
+    return (dispatch: Function) => dispatch(showErrorNotification(props));
```
Thanks, fixed it

zbettenbuk commented on react/features/recording/components/AbstractRecordingLabel.js:
```diff
@@ -34,8 +34,49 @@ export type Props = {
+     * @inheritdoc
+     */
+    render() {
+        if (!this.props._visible) {
```
Fixed, thanks!

zbettenbuk commented:
> There's also this warning. Do you have any idea what could be causing it ?

This comes from the Dialog.native.js, see note there

> * XXX The value may trigger a react-native warning in the Debug configuration
> * but, unfortunately, I couldn't find a value that wouldn't.

zbettenbuk commented:
> So it appears that this PR is not using the right config option for enabling the FILE recording. Could you please make sure to enable it on fileRecordingsEnabled equals true ? Currently on meet.jit.si it's disabled, but the mobile app turns it on for me.

Sure. That property appeared just recently due to a refactor. When I started implementing the feature, it didn't exist yet.

zbettenbuk commented:
> Also we should probably add "live streaming is starting..." and "live streaming stopped" notifications similar to what we have on web. Starting live streaming is long process and it's good to let the user know that someone is starting it. There are no sounds currently, so it can be a surprise.

Live streaming is my next task separate from this PR

paweldomas commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+    const { enabled, notifications } = state['features/notifications'];
+
+    return {
+        _notifications: notifications,
```
It is old code, but part of this PR according to @lyubomir was to write a solid base for notifications and "old code" can not be excuse. This flag is part of the abstraction.

I can also imagine use cases for this flag, but the questions is do they exist currently ? If not then I'd say it's better to remove it.

zbettenbuk replied:
Oh, I'm not reasoning not to remove, I just wanna know if it would break something or not. @virtuacoplenny do you know if that flag was necessary for AtlasKit?

zbettenbuk commented:
> As raised by Saul the calendar's ConferenceNotification needs to be unified
with the new notification abstraction added here. I'm not sure if this should be other type of notifications or do we want all of them to appear the same way. Until we have a discussion and make a decision it could be unified with an option which tells how notification is displayed and with this option calendar notifications would appear on the side as they currently do. When thinking about this I'm wondering what will happen if both calendar and recording notifications will appear at the same time ?

>Also I have a concern about enabling FILE recordings on mobile at this point given that the recording's file is not accessible yet. It would only confuse users. In addition to that I learned from Brian that the file is currently stored on Jibri and without the script which uploads and deletes it, it is possible to fill up Jibri machine. Can we add a flag which would be false by default and the recording menu item would stay hidden until the FILE recording is usable ?

> From testing on mobile:

> Jibri will appear in the filmstrip in the following scenario:

> Join room A and start the recording, wait for the recording to start.
>Leave the room A without stopping the recording.
> Join the room A again - Jibri will appear as one of the participants.
> Another time when the recording failed to start I ended up with 3 notifications:
> "Preparing to record the meeting" (with no way to dismiss)
recording failed (I don't have the exact message)
> "Recording stopped"
> Then the app disconnect, so I couldn't try to dismiss them, but I guess that at this point the "Preparing to record the meeting" would not disappear.

Thanks, I fixed the sticky notification when an error happens, also fixed the layout of bigger notifications you mentioned in an other comment, thanks for noticing!

About the calendar notification: I fully agree, I just don't think it should be part of this PR.

Regarding the jibri in the filmstrip: I can't reproduce that issue. Can you please give me some more details? Environment, device? Also, is this filtering done in Filmstrip?

zbettenbuk commented:
> Ok then I probably misunderstood this comment (I've got the impression it's part of this PR, because the labels work):

> Live streaming was not in my plans, but just got a question from Lyubo indicating that it'll be necessary too, so I'll revisit this

Right, I didn't comment it later on. Lyubo asked me to discuss it with the reviewers to leave Live streaming out of this PR because I think it's already too big. I'm already working on streaming, rebased to this branch, so it should come as a sequence, if you agree.

paweldomas commented:
> Right, I didn't comment it later on. Lyubo asked me to discuss it with the reviewers to leave Live streaming out of this PR because I think it's already too big. I'm already working on streaming, rebased to this branch, so it should come as a sequence, if you agree.

Yes, lets leave it for the next PR.

paweldomas commented:
>>Also I have a concern about enabling FILE recordings on mobile at this point given that the recording's file is not accessible yet. It would only confuse users. In addition to that I learned from Brian that the file is currently stored on Jibri and without the script which uploads and deletes it, it is possible to fill up Jibri machine. Can we add a flag which would be false by default and the recording menu item would stay hidden until the FILE recording is usable ?

> Well, my task was to implement it, I guess we can disable it by the config.js on LIVE until it's not needed (?). But doesn't the same limitation affect web?

My confusion came from the fact that the FILE recording was working on meet.jit.si and was not usable, but after changing the flag to `fileRecordingsEnabled` it's now all good in that topic, so forget about it.

zbettenbuk commented on react/features/notifications/components/NotificationsContainer.native.js:
```diff
@@ -0,0 +1,155 @@
+/**
+ * The height of the filmstrip in narrow aspect ratio.
+ */
+const FILMSTRIP_HEIGHT = 90;
```
Fixed, thanks for the suggestion!

zbettenbuk commented on react/features/notifications/components/NotificationsContainer.native.js:
```diff
@@ -0,0 +1,155 @@
+/**
+ * The height of the toolbox.
+ */
+const TOOLBOX_HEIGHT = 60;
```
Fixed, thanks for the suggestion!

zbettenbuk commented on react/features/notifications/components/AbstractNotificationsContainer.js:
```diff
@@ -0,0 +1,155 @@
+                    : undefined;
+
+            if (notification !== previousNotification) {
+                this._notificationDismissTimeout
```
Fixed, thanks!

zbettenbuk commented:
@paweldomas Pushed some new changes, refactored all the notification code that you suggested, I hope I didn't miss any of your points. Feel free to give it a final review, but you can focus on the notifications commit, because only that changed today.

paweldomas commented:
> I can't reproduce that issue. Can you please give me some more details? Environment, device? Also, is this filtering done in Filmstrip?

I can't reproduce it anymore either. Maybe one of the recent changes around poltergeist have fixed it. Lets just keep an eye on it.
